src/Sylius/Bundle/ThemeBundle/Templating/Locator/TemplateLocator.php

```diff
@@ -0,0 +1,271 @@
+     */
+    private $paths = [
+        'bundle_template' => [
+            '%theme_path%/%bundle_name%/%override_path%',
+            '%app_path%/Resources/%bundle_name%/%override_path%',
+            '%bundle_path%/Resources/%override_path%',
+        ],
+        'app_template' => [
+            '%theme_path%/%override_path%',
+            '%app_path%/Resources/%override_path%',
+        ],
+    ];
+
+    /**
+     * @param KernelInterface $kernel
+     * @param ThemeRepositoryInterface $themeRepository
+     * @param ThemeContextInterface $themeContext
+     * @param string $appDir
+     * @param string $cacheDir The cache path
+     */
+    public function __construct(
+        KernelInterface $kernel,
+        ThemeRepositoryInterface $themeRepository,
+        ThemeContextInterface $themeContext,
+        $appDir,
+        $cacheDir = null
+    ) {
+        $this->kernel = $kernel;
+        $this->themeRepository = $themeRepository;
+        $this->themeContext = $themeContext;
+        $this->appDir = $appDir;
+
+        if (null !== $cacheDir && is_file($cache = $cacheDir . '/templates.php')) {
+            $this->cache = require $cache;
+        }
+    }
+
+    /**
+     * Returns a full path for a given file.
+     *
+     * @param TemplateReferenceInterface $template A template
+     * @param string $currentPath Unused
+     * @param bool $first Unused
+     *
+     * @return string The full path for the file
+     *
+     * @throws \InvalidArgumentException When the template is not an instance of TemplateReferenceInterface
+     * @throws \InvalidArgumentException When the template file can not be found
+     */
+    public function locate($template, $currentPath = null, $first = true)
+    {
+        if (!$template instanceof TemplateReferenceInterface) {
+            throw new \InvalidArgumentException("The template must be an instance of TemplateReferenceInterface.");
+        }
+
+        $themes = $this->themeContext->getThemesSortedByPriorityInDescendingOrder();
+
+        $templatePath = $this->locateTemplateUsingThemes($template, $themes);
+
+        if (null === $templatePath) {
+            throw new \InvalidArgumentException(
+                sprintf('Unable to find template "%s" (using themes: %s).', $template, join(', ', $themes))
+            );
+        }
+
+        return $templatePath;
+    }
+
+    /**
+     * @param TemplateReferenceInterface $template
+     * @param ThemeInterface[] $themes
+     *
+     * @return null|string
+     */
+    protected function locateTemplateUsingThemes(TemplateReferenceInterface $template, array $themes)
+    {
+        foreach (array_merge($themes, [null]) as $theme) {
+            $result = $this->getCache($template, $theme);
+            if (null !== $result) {
+                return $result;
+            }
+        }
+
+        if (0 === strpos($template->getPath(), '@')) {
+            return $this->locateBundleTemplateUsingThemes($template, $themes);
+        } else {
```

Comment:
Sure, but I think that current version is cleaner, it's all about personal preferences. 

Right now, `else` block implies that these methods are bound up with each other. Without it, it would seem like locating app templates is some kind of fallback method.